Rename refactoring support for Rascal (locals)

toinehartman opened the pull request:
Add rename support for Rascal itself (but not for DSLs). Can be triggered in VS Code using F2 or right-click->rename.

The renaming feature is intended to be released incrementally. This PR includes milestones 1&2 as described in #382.

DavyLandman commented on rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc:
```diff
@@ -0,0 +1,275 @@
+                                             });
+}
+
+bool isImplicitDefinition(start[Module] _, rel[loc use, loc def] useDef, Define _: <_, _, _, variableId(), defined, _>) = defined in useDef.use; // use of name at implicit declaration loc
```
nit: I think instead of `Define _: <...>` we can just say  `<...>`

toinehartman replied:
I am aware of that, but was trying to prevent messing up the arity or types of the tuples here.

DavyLandman replied:
Fair, it's tricky sometimes.

DavyLandman commented on rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc:
```diff
@@ -0,0 +1,275 @@
+bool isImplicitDefinition(start[Module] _, rel[loc use, loc def] useDef, Define _: <_, _, _, variableId(), defined, _>) = defined in useDef.use; // use of name at implicit declaration loc
+
+bool isImplicitDefinition(start[Module] m, rel[loc, loc] _, Define _: <_, _, _, patternVariableId(), defined, _>) {
+    visit (m) {
```
I get what is happening here, but it feels like there should be a more rascal way? especially the if part of it. like, why it's not part of the pattern match in `qualifiedName` or something?

toinehartman replied:
Fixed.

DavyLandman commented on rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc:
```diff
@@ -0,0 +1,275 @@
+
+void checkCausesCapture(TModel tm, start[Module] m, set[Define] currentDefs, set[loc] currentUses, set[Define] newDefs) {
+    set[Define] newNameImplicitDefs = {def | Define def <- newDefs, isImplicitDefinition(m, tm.useDef, def)};
+    set[Define] newNameExplicitDefs = {def | Define def <- newDefs, !isImplicitDefinition(m, tm.useDef, def)};
```
nit: these two seem like set difference:

```
allDefs = {def | Define def <- newDefs };
newNameImplicitDefs = {d | d <- allDefs, isImplicitDefinition(m, tm.useDef, d)};
newNameExplicitDefs = allDefs - newNameImplicitDefs ;
```

and ideally (also faster):

```
newNameImplicitDefs = findImplicitDefs(m, tm.useDef); // only a single visit over the tree instead of a visit over the tree for every define
newNameExplicitDefs = {def | Define def <- newDefs } - newNameImplicitDefs;
```

toinehartman replied:
This is not purely a visit over the tree. In case of pattern variables, this syntactical analysis is needed, but for regular variables, it is not. However, I understand the point and can rewrite to something that visits the tree at most once.

DavyLandman commented on rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc:
```diff
@@ -0,0 +1,275 @@
+}
+
+void checkEditableDefinitions(set[Define] currentDefs) {
+    bool isEditableLocation(loc l) = l.scheme in {"file", "cwd", "home", "memory", "project"};
```
It's not possible to make an accurate list, and also, if a file is readonly on disk, we still cannot write to it. Let's let VS Code worry about updating files or not, as it's a whole mess in the virtual file system if something is readonly or not.

DavyLandman replied:
Yeah, I like the thought, but in the end, I don't think we can get this water tight, without adding a `isWritable(loc)` function to `IO.rsc` (not even sure if we can do that properly)

toinehartman replied:
Different approach (not for this PR, but for cross-module features that we're working on next): reject the renaming if any related definition is outside of the current workspace?

DavyLandman commented on rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc:
```diff
@@ -0,0 +1,275 @@
+        checkLegalName(newName);
+        checkCausesDoubleDeclaration(currentDefs, newNameDefs);
+        checkCausesCapture(tm, m, currentDefs, currentUses, newNameDefs);
+    } catch IllegalRenameReason r: {
```
I'm not sure I truly like this style of using exceptions for "regular" control flow. I get the beautify, but there are some constrains to this style:

1. it only allows for a single message
2. we have to be careful about other exceptions.

if every checker would return a set of `IllegalRenameReason` we can plus all the calls together, and if the result is non-empty, throw it is an illegal rename. It might also remove the need for `throwIfNotEmpty` as we can use the comprehension directly (with a small change of the signature of the `IllegalRenameReason` ADT I guess).

DavyLandman commented on rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc:
```diff
@@ -0,0 +1,275 @@
+        when moduleLoc == scope; // function is defined in module scope
+    default Maybe[str] isUnsupportedDefinition(Define _) = nothing();
+
+    throwIfNotEmpty(UnsupportedRename, {<def.defined, msg> | def <- defsToRename, just(msg) := isUnsupportedDefinition(def)});
```
if we switch to returning a set, we can get rid of quite some maybe's as well.

DavyLandman commented on rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc:
```diff
@@ -0,0 +1,275 @@
+void checkUnsupported(loc moduleLoc, set[Define] defsToRename) {
+    Maybe[str] isUnsupportedDefinition(Define _: <scope, id, _, functionId(), _, _>) = just("Global function definitions might span multiple modules; unsupported for now.")
+        when moduleLoc == scope; // function is defined in module scope
+    default Maybe[str] isUnsupportedDefinition(Define _) = nothing();
```
wouldn't it be easier to mark what we do support? instead of all the cases we don't? Like I'm not seeing module names or grammar names for example.

toinehartman replied:
In geneal, `findDeclarationName` will take care of what we do and don't support. In this case however, we need some more analysis to see if we really support this definition.

toinehartman replied:
Done!

DavyLandman commented on rascal-lsp/src/main/rascal/lang/rascal/tests/Rename.rsc:
```diff
@@ -0,0 +1,410 @@
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+module lang::rascal::tests::Rename
```
🎉 nice that we have tests! 🎉

DavyLandman commented on rascal-vscode-extension/test-workspace/test-project/src/main/rascal/RenameExamples.rsc:
```diff
@@ -0,0 +1,74 @@
+module RenameExamples
```
do we still need this file as well? and if so, add a small comment on top of why.

DavyLandman commented on rascal-vscode-extension/test-workspace/test-project/src/main/rascal/SingleModuleRenameTest.rsc:
```diff
@@ -0,0 +1,31 @@
+module SingleModuleRenameTest
```
similar for this file, do we need it? or can we move it to the `rascal-lsp/test` directory?

DavyLandman reviewed (changes requested):
All in all, I'm quite happy, looks good, this has been a significant amount of work. I think my main question is regarding the exceptions for the errors.

And it might be nice to have 1 integration test added to the UI test to make sure the rename also works (and keeps working).

toinehartman commented on rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java:
```diff
@@ -114,11 +120,13 @@ public class RascalTextDocumentService implements IBaseTextDocumentService, Lang
         this.ownExecuter = exec;
         this.documents = new ConcurrentHashMap<>();
         this.columns = new ColumnMaps(this::getContents);
+        this.docChanges = new DocumentChanges(this);
```
That has been on my list for a while. Now done in f141a96f.

toinehartman commented on rascal-lsp/pom.xml:
```diff
@@ -160,6 +174,13 @@
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-surefire-plugin</artifactId>
         <version>${maven-surefire-plugin.version}</version>
+        <configuration>
```
Turns out not to be necessary. Removed.

toinehartman commented:
> All in all, I'm quite happy, looks good, this has been a significant amount of work. I think my main question is regarding the exceptions for the errors.
> 
> And it might be nice to have 1 integration test added to the UI test to make sure the rename also works (and keeps working).

Thanks. Processed all your remarks and added an integration test. Can you please have another look?

DavyLandman commented on rascal-vscode-extension/src/test/vscode-suite/ide.test.ts:
```diff
@@ -152,5 +153,21 @@ describe('IDE', function () {
+        await editor.moveCursor(7, 15);
+
+        await bench.executeCommand("Rename Symbol");
+        const renameBox = await driver.wait(async () => await bench.getEditorView().findElement(By.className("rename-input")), Delays.fast, "Rename box should appear");
```
maybe we can do something that reuses the same functions that are present in `util.ts`?

toinehartman replied:
Had a look, but I don't see anything related to finding arbitrary editor elements there. I can refactor once we write more related tests (e.g. cross-module and/or other refactorings).

DavyLandman replied:
I meant that for example the `hasElement` function looks to do exactly what this line is doing.

toinehartman replied:
Missed that completely. Fixed it.

DavyLandman reviewed (approved):
Looks good to me. We'll have to wait for the checker integration to finish, and then we can finish this PR.

jurgenvinju commented on rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc:
```diff
@@ -0,0 +1,272 @@
+
+    // Will this rename turn a use of <oldName> into a use of an existing declaration of <newName> (shadowing)?
+    rel[loc use, loc oldDef, loc newDef] explicitDeclShadowsCurrentDecl =
+        {<cU, cD.defined, nD.defined> | Define nD <- newNameExplicitDefs
```
if this is a efficiency bottleneck then rewriting a bit with the `o` composition operator may take off the edge. It typically avoids one more iteration through the target set, so takes off one complexity order. I think there are two hidden `o` operators here before two filters.

toinehartman replied:
I will think about that. Performance has not been my primary concern and since this should only consider local variables (otherwise an exception will be thrown before this code is even called), I am not too worried right now.

toinehartman replied:
Actually went ahead and rewrote this. Much better, thanks!

jurgenvinju reviewed (approved):
very nice. can't wait to see it in action :-)

toinehartman commented on rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc:
```diff
@@ -0,0 +1,275 @@
+    | UnexpectedFailure(str message)
+;
+
+private set[loc] getUses(TModel tm, loc def) {
```
Yeah, they are all at the bottom (and marked with comments). I am not sure if these will be exposed in the new typechecker. Will have a look once we get there.

toinehartman commented:
I pushed some more changes, addressing some remarks by @PaulKlint outside of this PR. Especially 5fe2162 should highly reduce the number of parse tree visits in real scenarios. Now, a module tree will be visited at most once.

toinehartman commented:
TODO:

- [ ] Keyword parameters (local functions only) (ht @sungshik)

DavyLandman commented on rascal-vscode-extension/src/test/vscode-suite/ide.test.ts:
```diff
@@ -155,5 +156,21 @@ describe('IDE', function () {
+
+        await bench.executeCommand("Rename Symbol");
+        const renameBox = await ide.hasElement(editor, By.className("rename-input"), Delays.fast, "Rename box should appear");
+        await renameBox.sendKeys(Key.BACK_SPACE, "i", Key.ENTER);
```
can we not use `setText` on the input box?

toinehartman replied:
As discussed in private, that's not straightforward here. Resolving.

DavyLandman commented on rascal-vscode-extension/src/test/vscode-suite/ide.test.ts:
```diff
@@ -155,5 +156,21 @@ describe('IDE', function () {
+        await bench.executeCommand("Rename Symbol");
+        const renameBox = await ide.hasElement(editor, By.className("rename-input"), Delays.fast, "Rename box should appear");
+        await renameBox.sendKeys(Key.BACK_SPACE, "i", Key.ENTER);
+        await driver.wait(async () => (await editor.isDirty()), Delays.extremelySlow, "Rename should have resulted in changes in the editor");
```
nit : `async () => (await editor.isDirty()` => `() => editor.isDirty()`

toinehartman replied:
Fixed

DavyLandman reviewed (approved):
Looks good, 2 questions on the test
